### ISSUE TITLE
Minimalist code generation for simple lambdas.

### ISSUE THREAD

**Issue by ericmutta** [dotnet/vblang#131]
Passing in functions/subs using lambdas is a very handy abstraction tool and will be used heavily in any sizable project. A simple example:

```vb
  Public Sub Foo()
    Dim data = New List(Of Integer) From {1, 2, 3}
    data.ForEach(Sub(n) Console.WriteLine(n))
  End Sub
```

The lambda sub passed to the `List.ForEach()` method makes NO REFERENCE to any members in the containing `Foo()` method however the compiler goes ahead and GENERATES A FULL CLASS to implement it as a closure, as follows:

```vb
Friend NotInheritable Class _Closure$__
    Friend Sub _Lambda$__0-0(ByVal n As Integer)
        Console.WriteLine(n)
    End Sub

    Public Shared ReadOnly $I As _Closure$__ = New _Closure$__
    Public Shared $I0-0 As Action(Of Integer)
End Class
```

What's worse is it does this for every single little lambda sub/function you use (it has no de-duplication if the lambdas have identical bodies) and pretty soon these automatically generated closure classes add up, adding overhead to the assembly (in my current project I have over 45 such closure classes because I use lambdas heavily).

In the simple case (where a lambda depends purely on its arguments and does not touch the enclosing environment) the compiler should generate minimal code consisting of just one function. The previous example should ideally be implemented as follows:

```vb
  Public Sub Foo()
    Dim data = New List(Of Integer) From {1, 2, 3}
    data.ForEach(AddressOf automatically_generated)
  End Sub
  Private Sub automatically_generated(n As Integer)
    Console.WriteLine(n)
  End Sub
```

**Comment by AnthonyDGreen** [comment]
The compiler generates a singleton closure object so that the lambda may be an instant method. Because of events (I imagine), the CLR has optimized code paths for methods that have an instance parameter. The performance difference is significant enough that initially we tried just generating every lambda they would otherwise be a Shared method with an extra dummy parameter. This however caused breaks in people's code because reflection now reported the extra parameter so all the parameter counts were off. As a mitigation we decided to instead of generating Shared methods with an extra parameter to use an instance method on a singleton class. Reflection doesn't report the instance parameter so only those parameters declared by the user appear _and_ lambdas such as yours which don't captured any state no longer suffer a performance penalty.

Does that make sense?

**Comment by franzalex** [comment]
@AnthonyDGreen No, it still doesn't make sense to me. What is the need to add the extra parameter for if the lambda expression is a simple one that references no external objects?

I can understand that it might probably be too much asking for lambdas with identical bodies pointing to a single closure due to some constraints in the CLR. However, with regards to the real issue that @ericmutta points out, simplifying a lambda that doesn't reference any members of the enclosing method into a method within the class will be a very welcome optimisation.

I believe it's one thing that could be easily implemented which will most likely save a few memory allocations and method calls.

**Comment by ericmutta** [comment]
@AnthonyDGreen many thanks for responding quickly!

So if I understand you correctly, the way the compiler generates an entire class right now is a performance optimisation because of how the CLR handles the resulting code...OK I can live with that (i.e the fact that there's a rationale for generating the code that way).

Along these lines of performance: would the compiler's code be more efficient than the handwritten equivalent where you use a function and AddressOf instead of a lambda? As in this code:

```vb
  Public Sub Foo()
    Dim data = New List(Of Integer) From {1, 2, 3}
    data.ForEach(AddressOf use_normal_function_instead_of_lambda)
  End Sub
  Private Sub use_normal_function_instead_of_lambda(n As Integer)
    Console.WriteLine(n)
  End Sub
```
I ask because that is the kind of code I am writing to avoid generation of hundreds of those closure classes in a large project...am I losing that performance gained by the way the compiler currently generates the code?

**Comment by AnthonyDGreen** [comment]
@VSadov, @AlekseyTs, @gafter have more context on this optimization and codegen in general. One difference I can imagine with your approach is that because instance methods are ... instance methods, two delegates created from two different objects won't be value equal, which causes some problems with removing delegates. By making them instance methods on a singleton all delegates pointing to that method are value equal (same method and same receiver). I don't remember if that specifically was why we chose this implementation.

With any performance motivated request it's best to first start by profiling your project to see where the bottlenecks are.

There are always trade-offs between assembly size in memory vs heap allocations vs execution time and there's no one answer that applies to all situations. I know there was a request on the C# side to cache explicit delegate creations like you're using similar to how we do lambdas to avoid memory allocations but in that instance we'd be adding to size on disk/memory to avoid some garbage at runtime. I can't even remember whether we decided to take that optimization.

**Comment by VSadov** [comment]
As with many things about performance the most definitive info would be from measuring the metrics that you care about.
Generally it is more common for programs to value per-invocation throughput improvement more than the size of metadata increase that is a relatively small expense in this case and has effect only once per-type load.

If you use delegates vs. lambdas, you will be doing different kinds of tradeoffs. Whether that is profitable, depends on your preferences - ie throughput vs. metadata size.

**Comment by ericmutta** [comment]
@AnthonyDGreen @VSadov thanks for the clarifications, I see now there are tradeoffs and will leave it up to you guys to decide whether this request should be implemented.

As a side note, my situation makes **both** throughput and metadata size important. I have a portable VB library that runs both on the server side (where throughput is important) and on the client-side where assembly size is important. The client side portion of the (Android) app has to frequently download new versions of the assembly over slow/unreliable 2G links, and with hundreds of thousands of clients, shaving off a few bytes really adds up!
